### ISSUE TITLE
Truthiness narrowing loses `Self` specialization on inherited attributes

### ISSUE THREAD

**Issue by sunny-zuo** [astral-sh/ty#3755]
### Summary

Starting in ty 0.0.45, truthiness narrowing of `Child | None` causes an inherited attribute annotated with `Self` to lose its subclass specialization. After narrowing to `Child`, the attribute is inferred as `list[Self@Base]` instead of `list[Child]`. This checked correctly in ty 0.0.44.

## Minimal reproduction ([playground](https://play.ty.dev/ae5dcc10-4671-4182-90d2-8c6b7227c97f))

```python
from typing import Self


class Base:
    values: list[Self]


class Child(Base):
    pass


def f(value: Child | None) -> list[Child]:
    if value:
        reveal_type(value.values)  # list[Self@Base] on 0.0.45, list[Child] on 0.0.44

        # Return type does not match returned value: expected `list[Child]`, found `list[Self@Base]`
        return value.values
    raise ValueError
```

### Version

ty 0.0.49

**Comment by charliermarsh** [comment]
This regressed in https://github.com/astral-sh/ruff/pull/25626, though unclear if we should revert, or fix narrowly, or wait for a larger reconsideration of `Self` in intersections (I think @sharkdp was perhaps going to look at this?).

**Comment by sharkdp** [comment]
I think there is a more fundamental problem here that we should discuss first before we solve any problems specific to intersections.

The questions is: what does this even mean?
```py
class Base:
    values: list[Self]
```
If you interpret `Self` as a covariant type variable, like @sunny-zuo's post implies ("expected `list[Child]`"), that leads to unsoundness. Given a `Base` instance, one could add a `Base` instance to that list, and so it's not guaranteed that all elements of `Child.values` are of type `Child`:
```py
class Base:
    values: list[Self]

def add_base(b: Base):
    b.values.append(Base())

class Child(Base):
    pass

c = Child()
add_base(c)
reveal_type(c.values)  # all type checkers, including ty: list[Child], oops
```

In https://discuss.python.org/t/unsoundness-of-contravariant-self-type/86338, @carljm argues that the only sound interpretation of a class-scoped `Self` is that it's a reference to `Base` itself, essentially just an alias for the class name. However, it seems that we have never followed up on that in ty?

This discussion is relevant here because I'm not even sure if `list[Child]` is the right answer. Do we want to accept this unsoundness? This is not related to `list[Self]`, by the way. The same problem appears for a simple mutable `value: Self` attribute.

**Comment by carljm** [comment]
Yes, I'm not clear on what we should do here. I think basically all users of mutable `Self` typed attributes are expecting the unsound covariant version (because that's what all other type checkers do) and many may be unaware of how unsound it is. This leads me to suspect that our best option is to continue implementing these in the unsound way, to meet user expectations, and perhaps consider a warning diagnostic on any such use? I don't think the "just an alias for the class name" use case is really compelling enough to implement (especially given Python 3.14+ annotation name lookup rules, which allow using the class name as a forward reference without quoting). So basically I think there is no use case for Self typed attributes that is both useful and sound, so the best we can do is support the "attractive-but-unsound" use case that everyone expects, and maybe warn about it to alert those who care.
